18F/identity-idp
# Wait for DOMContentLoaded to initialize React document capture

**Why:** To improve interoperability with default application scripts which also use DOMContentLoaded to apply default behavior to server-rendered content. As long as we can assume that the document capture script is loaded after the default application script, it can be assured that DOMContentLoaded event handlers will occur only after default scripts are applied, where said previous scripts can safely operate under the assumption that content on the page at that moment in time is rendered by the server, and not by other JavaScript (the React application).

This resolves an issue where the React-rendered form was being subjected to application-wide submit button auto-enabling behavior. Content rendered by React should be self-contained.

See: https://github.com/18F/identity-idp/blob/master/app/javascript/app/form-validation.js

Included are additional unit tests which, while not directly relevant to the fix, are included since they improve test coverage. They were originally authored as part of a TDD(-ish) approach to assuming the bug originated in the React application. See also first note in "Alternatives Considered".

**Alternatives Considered:**

- There should be no default behavior, because it is difficult to debug if one doesn't have awareness that these default script behaviors exist in the first place. Instead, all behaviors should be explicit opt-in.
- Allow React application to _opt out_. I originally approached it this way with the `no-auto-enable` button class. However, there are still some default behaviors which the React form would have been subjected to. I opted instead to avoid those altogether. Instead, the default behavior could disable itself for e.g. an entire form, or an entire page.
- Following the logic of the default behavior, the React form should apply [`required` HTML attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/required) to image inputs even if the form progression is managed automatically. This might be good practice to implement, even if it's still desired that the React form still not be subject to the default auto-enable behavior.

**Testing Instructions:**

1. (Prerequisite) Ensure `document_capture_step_enabled: 'true'` in `config/application.yml`
2. Log in and navigate to `/verify`
3. Fill in a value for "Front Image"
4. "Continue" button should remain disabled
